[vulkan] Trigger global Vulkan runtime to load asynchronously on startup/library load

SS-JIA opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#76770 [vulkan] Trigger global Vulkan runtime to load asynchronously on startup/library load**

## Context 

This PR modifies the behaviour of the Vulkan backend to trigger asynchronous initialization of the global Vulkan runtime on application load or library load for Android devices. The first time a Vulkan model is loaded, the global Vulkan runtime will be initialized, which involves loading Vulkan drivers and creating a logical device. This drastically increases the model load latency.

However, the Vulkan runtime can be pre-emptively intialized on application load/library load so that it has no effect on model load latency. This is done by triggering the initialization via a static variable.

Runtime initialization will only load Vulkan drivers and create a logical Vulkan device; it will not allocate GPU memory, etc. Therefore the impact of pre-emptive initialization is minimized if the application does not end up using any Vulkan models.

## Benchmark Results

Model load time tests with `model_load_benchmark` using a person segmentation model, on a Pixel 5:

CPU Baseline: 41 ms
Vulkan Baseline: 72 ms
Vulkan Baseline, with pre-emptive loading: 44 ms

Differential Revision: [D36111656](https://our.internmc.facebook.com/intern/diff/D36111656)

Commit:
[vulkan] Trigger global Vulkan runtime to load asynchronously on startup/library load

[ghstack-poisoned]

Commit:
Update on "[vulkan] Trigger global Vulkan runtime to load asynchronously on startup/library load"


## Context 

This PR modifies the behaviour of the Vulkan backend to trigger asynchronous initialization of the global Vulkan runtime on application load or library load for Android devices. The first time a Vulkan model is loaded, the global Vulkan runtime will be initialized, which involves loading Vulkan drivers and creating a logical device. This drastically increases the model load latency.

However, the Vulkan runtime can be pre-emptively intialized on application load/library load so that it has no effect on model load latency. This is done by triggering the initialization via a static variable.

Runtime initialization will only load Vulkan drivers and create a logical Vulkan device; it will not allocate GPU memory, etc. Therefore the impact of pre-emptive initialization is minimized if the application does not end up using any Vulkan models.

## Benchmark Results

Model load time tests with `model_load_benchmark` using a person segmentation model, on a Pixel 5:

CPU Baseline: 41 ms
Vulkan Baseline: 72 ms
Vulkan Baseline, with pre-emptive loading: 44 ms



[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/vulkan/api/Runtime.cpp:
```diff
@@ -274,6 +287,29 @@ std::unique_ptr<Runtime> init_global_vulkan_runtime() {
+  // This is safe since the global vulkan runtime instance is stored as a static
+  // local variable inside runtime(). Local static variable initialization is
+  // thread-safe since C++11.
+  std::thread async_loader(runtime);
```
My worry is that you can still call `init_global_vulkan_runtime` from multiple threads and each of those can create an instance of Runtime. Only one of them will succeed in constructing `unique_ptr<Runtime>`. How can we validate this? 

However, you are still tying it to library load time so it probably is ok as you wont have instances where more than one thread is executing runtime.

SS-JIA commented:
@SS-JIA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36111656).

Commit:
Update on "[vulkan] Trigger global Vulkan runtime to load asynchronously on startup/library load"


## Context 

This PR modifies the behaviour of the Vulkan backend to trigger asynchronous initialization of the global Vulkan runtime on application load or library load for Android devices. The first time a Vulkan model is loaded, the global Vulkan runtime will be initialized, which involves loading Vulkan drivers and creating a logical device. This drastically increases the model load latency.

However, the Vulkan runtime can be pre-emptively intialized on application load/library load so that it has no effect on model load latency. This is done by triggering the initialization via a static variable.

Runtime initialization will only load Vulkan drivers and create a logical Vulkan device; it will not allocate GPU memory, etc. Therefore the impact of pre-emptive initialization is minimized if the application does not end up using any Vulkan models.

## Benchmark Results

Model load time tests with `model_load_benchmark` using a person segmentation model, on a Pixel 5:

CPU Baseline: 41 ms
Vulkan Baseline: 72 ms
Vulkan Baseline, with pre-emptive loading: 44 ms

Differential Revision: [D36111656](https://our.internmc.facebook.com/intern/diff/D36111656)

[ghstack-poisoned]

SS-JIA commented:
@SS-JIA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36111656).
